File: backend-server/app/Http/Listeners/IngestCodebaseDataSource.php
<?php

namespace App\Http\Listeners;

use App\Http\Enums\IngestStatusType;
use App\Http\Events\CodebaseDataSourceWasAdded;
use App\Models\CodebaseDataSource;
use Exception;
use GuzzleHttp\Client;
use GuzzleHttp\Exception\GuzzleException;
use Illuminate\Contracts\Queue\ShouldQueue;

class IngestCodebaseDataSource implements ShouldQueue
{

    /**
     * @throws GuzzleException
     * @throws Exception
     */
    public function handle($event)
    {
        if (!$event instanceof CodebaseDataSourceWasAdded) {
            return;
        }

        $botId = $event->getChatbotId();
        $codebaseDataSourceId = $event->getCodebaseDataSourceId();


        /** @var CodebaseDataSource $datasource */
        $datasource = CodebaseDataSource::find($codebaseDataSourceId);


        $repo = $datasource->getRepository();

        $requestBody = [
            'type' => 'codebase',
            'repo' => $repo,
            'namespace' => $botId,
        ];

        // Call to ingest service endpoint
        $client = new Client();
        $response = $client->request('POST', 'http://llm-server:3000/api/ingest', [
            'json' => $requestBody,
        ]);

        $datasource->setIngestedAt(now());

        if ($response->getStatusCode() !== 200) {
            $datasource->setIngestionStatus(IngestStatusType::FAILED);
        } else {
            $datasource->setIngestionStatus(IngestStatusType::SUCCESS);
        }

        $datasource->save();
    }
}

File: dj_backend_server/web/listeners/ingest_codebase_data_source.py
# listeners.py
import os
import requests
from requests.exceptions import RequestException
from web.models.codebase_data_sources import CodebaseDataSource
from web.enums.ingest_status_enum import IngestStatusType
from web.signals.codebase_datasource_was_created import codebase_data_source_added
from django.core.exceptions import ObjectDoesNotExist
from django.utils.timezone import now
from web.signals.codebase_datasource_was_created import codebase_data_source_added

@codebase_data_source_added.connect
def ingest_codebase_data_source(sender, chatbot_id, data_source_id, **kwargs):
    try:
        datasource = CodebaseDataSource.objects.get(id=data_source_id)
    except ObjectDoesNotExist:
        return

    repo = datasource.repository

    request_body = {
        'type': 'codebase',
        'repo': repo,
        'namespace': str(chatbot_id),
    }

    try:
        # Call to ingest service endpoint
        url = os.getenv('APP_URL') + "/api/ingest/" 
        response = requests.post(url, json=request_body)

        datasource.ingested_at = now()

        if response.status_code != 200:
            datasource.ingestion_status = IngestStatusType.FAILED
        else:
            datasource.ingestion_status = IngestStatusType.SUCCESS

        datasource.save()

    except RequestException as e:
        datasource.ingested_at = now()
        datasource.ingestion_status = IngestStatusType.FAILED
        datasource.save()
        return